docs/notebooks/code_overview.pct.py

```diff
@@ -0,0 +1,238 @@
+# %% [markdown]
+# # An overview of Trieste types
+
+# %% [markdown]
+# Trieste is dedicated to Bayesian optimization, the process of finding the *optimal values of an expensive, black-box objective function by employing probabilistic models over observations*. This notebook explains how the different parts of this process are represented by different types in the code, and how these types can be extended.
+
+# %% [markdown]
+# ## Key types
+#
+# The following types represent the key concepts in Trieste. For a full listing of all the types in Trieste, see the API reference.
+
+# %% [markdown]
+# ### `Observer`
+#
+# The `Observer` type definition represents the black box objective function. Observers are functions that accept query points and return observations. Observations are either a single objective value that we wish to optimize, or a dictionary of mutiple tagged values that must be combined somehow, for example an objective and an inequality constraint. Objective values can be either single or multi-dimensional (see [multi-objective optimization](multi_objective_ehvi.ipynb)).
+
+# %% [markdown]
+# ### `Dataset`
+#
+# The `Dataset` container class represents the query points and observations from a single observer. Observers with multiple observations are represented by a dictionary of multiple tagged `Dataset`s.
+
+# %% [markdown]
+# ### `ProbabilisticModel`
+#
+# The `ProbabilisticModel` protocol represents any probabilistic model used to model observations. Like for `Dataset`, observers with multiple observations are modelled by a dictionary o multiple tagged models.
+#
+# At it simplest, a `ProbabilisticModel` is anything that implements a `predict` and `sample` method. However, many algorithms in Trieste depend on models with additional features, which are represented by the various subclasses of `ProbabilisticModel`. The standard Bayesian optimizer uses `TrainableProbabilisticModel` models, which also implement `update` and `optimize` methods. Specific acuqisition functions may require other features, represented by classes like `SupportsPredictJoint` and `SupportsGetObservationNoise`. Since these are defined as protocols, it is possible to define and depend on the intersections of different model types (e.g. only support models that are both `SupportsPredictJoint` and `SupportsGetObservationNoise`).
```

Comment:
It would be good to explain `update` and `optimize`. Maybe all in brackets?  

(...) which also implement `update` (updating the model structure when new data is added to the training set) and `optimize` (optimizing the model training loss) methods

also 
`SupportsPredictJoint` (ability to predict the joint distribution at several inputs x1, ... xp)
`SupportsGetObservationNoise` (ability to predict the observation noise variance)